consensus/src/consensus_agent/sync.rs

```diff
@@ -1,72 +1,340 @@
-    type Block: Serialize + Deserialize;
+#[derive(Debug, PartialEq, Eq)]
+pub enum SyncEvent<BE: BlockError> {
+    BlockProcessed(Blake2bHash, Result<PushResult, PushError<BE>>),
+}
 
-    fn get_block_locators(&self, max_count: usize) -> Vec<Blake2bHash>;
-    fn request_block(&self, block_hash: Blake2bHash) -> Self::Block;
+pub struct FullSync<'env, B: AbstractBlockchain<'env>> {
+    blockchain: Arc<B>,
+    peer: Arc<Peer>,
+    notifier: RwLock<PassThroughNotifier<'static, SyncEvent<<B::Block as Block>::Error>>>,
 }
 
+impl<'env, B: AbstractBlockchain<'env>> SyncProtocol<'env, B> for FullSync<'env, B> {
+    fn new(blockchain: Arc<B>, peer: Arc<Peer>) -> Arc<Self> {
+        Arc::new(Self {
+            blockchain,
+            peer,
+            notifier: RwLock::new(PassThroughNotifier::new()),
+        })
+    }
 
+    fn get_block_locators(&self, max_count: usize) -> Vec<Blake2bHash> {
+        self.blockchain.get_block_locators(max_count)
+    }
 
+    fn request_blocks(&self, locators: Vec<Blake2bHash>, max_results: u16) {
+        self.peer.channel.send_or_close(GetBlocksMessage::new(
+            locators,
+            max_results,
+            GetBlocksDirection::Forward,
+        ));
+    }
+
+    fn on_block(&self, block: B::Block) {
+        let hash = block.hash();
+        let result = self.blockchain.push(block);
+        self.notifier.read().notify(SyncEvent::BlockProcessed(hash, result));
+    }
+
+    fn on_epoch_transactions(&self, _epoch_transactions: EpochTransactionsMessage) {
+        warn!("We didn't expect any epoch transactions from {} - discarding and closing the channel", self.peer.peer_address());
+        self.peer.channel.close(CloseType::UnexpectedEpochTransactions);
+    }
 
+    fn register_listener<L: PassThroughListener<SyncEvent<<B::Block as Block>::Error>> + 'static>(&self, listener: L) {
+        self.notifier.write().register(listener)
+    }
 
-struct FullSync<B: AbstractBlockchain<'static>> {
-    blockchain: Arc<B>
+    fn deregister_listener(&self) {
+        self.notifier.write().deregister()
+    }
 }
 
-impl<B: AbstractBlockchain<'static>> FullSync<B> {
-    fn new(blockchain: Arc<B>) -> Self {
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum MacroBlockSyncingState {
+    MacroBlocks,
+    MicroBlocks,
+    Finished,
+}
+
+struct MacroBlockSyncState {
+    /// Cache of the next max. 1000 blocks.
+    block_cache: VecDeque<AlbatrossBlock>,
+    /// Transactions of the current block.
+    transactions_cache: Vec<Transaction>,
+    /// The current state of the syncing.
+    state: MacroBlockSyncingState,
+    /// Boolean flag whether we are currently processing an epoch.
+    processing_epoch: bool,
+}
+
+impl MacroBlockSyncState {
+    fn new() -> Self {
         Self {
-            blockchain
+            block_cache: VecDeque::new(),
+            transactions_cache: Vec::new(),
+            state: MacroBlockSyncingState::Finished,
+            processing_epoch: false,
         }
     }
 }
```

Comment:
Is there any chance this will have arguments in the future? If not, derive `Default` for the struct and just use that. Also clippy will warn, if you have a constructor without arguments, but no `Default`.